### ISSUE TITLE
devtools memory benchmarks not reporting correct memory numbers?

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#97588]
SkiaPerf flagged that the `maxRss` and `maxAdbTotal` metrics of the `complex_layout_scroll_perf__devtools_memory` on Android with a Linux host went from very large to very small numbers

https://flutter-flutter-perf.skia.org/e/?begin=1642621504&end=1643716522&keys=X74ea72abb6e43e95eb0f932f96f17d6a&requestType=0&xbaroffset=27346

On the Engine roll here: https://github.com/flutter/flutter/pull/97445.

However, I don't see any changes that could be responsible for such a dramatic drop in memory. I suspect the benchmark is not reporting the correct numbers. Looking over the changes in the roll, I'm not able to eyeball a culprit, but since these are `devtools` benchmarks, that's where I'll start.

/cc @kenzieschmoll @jason-simmons @iskakaushik

**Comment by jason-simmons** [comment]
This benchmark is using `flutter drive --profile-memory` which uses the DevTools `record-memory-profile` flag.

After the https://github.com/flutter/flutter/pull/97445 engine roll, if I cd to `dev/benchmarks/complex_layout` and run:
`flutter drive --profile --profile-memory devtools_memory.json --no-publish-port -v test_driver/scroll_perf.dart`

it produces this exception:
```
VMServiceFlutterDriver: Connecting to Flutter application at
http://127.0.0.1:62868/oJFA3Melbx8=/
VMServiceFlutterDriver: Isolate found with number: 1668169569313903
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
Unhandled exception:
LateInitializationError: Field 'eventSample' has not been
initialized.
#0      MemoryProfile.eventSample
(package:dds/src/devtools/memory_profile.dart)
#1      MemoryProfile._recalculate
           (package:dds/src/devtools/memory_profile.dart:309:7)
#2      MemoryProfile._updateGCEvent
           (package:dds/src/devtools/memory_profile.dart:278:5)
#3      MemoryProfile._handleGCEvent
           (package:dds/src/devtools/memory_profile.dart:146:5)
#4      _RootZone.runUnaryGuarded (dart:async/zone.dart:1618:10)
#5      _BufferingStreamSubscription._sendData
           (dart:async/stream_impl.dart:341:11)
#6      _DelayedData.perform (dart:async/stream_impl.dart:591:14)
#7      _StreamImplEvents.handleNext
(dart:async/stream_impl.dart:706:11)
#8      _PendingEvents.schedule.<anonymous closure>
           (dart:async/stream_impl.dart:663:7)
#9      _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#10     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#11     _runPendingImmediateCallback
           (dart:isolate-patch/isolate_patch.dart:122:13)
#12     _RawReceivePortImpl._handleMessage
           (dart:isolate-patch/isolate_patch.dart:193:5).
```

https://github.com/dart-lang/sdk/commit/d1baf327a18dfc92476049a09985b645c08072f5 was in that engine roll - possibly that is related?

@bkonyi

**Comment by bkonyi** [comment]
I doubt it's dart-lang/sdk@d1baf32, but it's almost certainly related to dart-lang/sdk@1c571a12fa1d4b13b363546360dbf5ff765bdf4e. I'll take a look.

**Comment by zanderso** [comment]
Thanks @jason-simmons. The same crash is in the logs from CI:

https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20complex_layout_scroll_perf__devtools_memory/1650/overview

Why didn't that crash cause the run to fail?

**Comment by jason-simmons** [comment]
That exception is logged by a DevTools process spawned by `flutter drive`:

`dart devtools --no-launch-browser --vm-uri=[URI] --record-memory-profile=devtools_memory.json`

But flutter_tools does not check the exit code of that process:
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/devtools_launcher.dart#L90

The failed DevTools run writes a `record-memory-profile` output file containing no data values.  The benchmark then reads it and produces inaccurate results.

**Comment by bkonyi** [comment]
Simple fix [here](https://dart-review.googlesource.com/c/sdk/+/231180).

**Comment by zanderso** [comment]
@kenzieschmoll Is there any reason the tool should not also exit with an error if that subprocess exits abnormally?

**Comment by kenzieschmoll** [comment]
In the current state, I believe we fail gracefully if the devtools process exits with an error. That way a user is not blocked from being able to launch their app due to a devtools related failure. Would having Flutter tool exit with an error prohibit this graceful failure?

**Comment by zanderso** [comment]
@kenzieschmoll On CI, if a test is expecting DevTools to be working correctly, and DevTools does not work correctly, then we'd like to have some way for the test to fail. The tool can detect that it's running on CI in various ways, for example by checking for environment variables, if that helps.

**Comment by kenzieschmoll** [comment]
Gotcha. If we exit only on the CI, I think that's reasonable.

**Comment by jason-simmons** [comment]
After that fix landed we started seeing `maxAdbTotal` metrics in this benchmark that still did not accurately reflect the memory consumption of the app under test.

(example: https://flutter-flutter-perf.skia.org/e/?begin=1643233812&end=1643929514&keys=Xaac119bd536e4a533973743df6e4fd5c&requestType=0&xbaroffset=27399)

Tried running `flutter drive --profile-memory` and noticed that the `adb_memoryInfo` entries in the output JSON file always contained the same values and did not update as the test continued to run.

Traced this to a null safety exception in Dart DDS.

Proposed fix: https://dart-review.googlesource.com/c/sdk/+/231533
